Add wolfSSL debug messages

gojimmypi opened the pull request:
# Description

The PR adds debug messages in various functions, mostly surrounding more informative failures. 

I found these useful while troubleshooting memory-constrained devices. It's particularly helpful to see the sizes of heap allocation attempts.

Of particular interest is [wolfSSL_new](https://github.com/gojimmypi/wolfssl/blob/2d7d68dc46ad43ca607e3c67cca9286bff9a3188/src/ssl.c#L1470C10-L1470C21) `XMALLOC` and the subsequent check of `InitSSL()` for positive return values. I didn't actually _see_ any. They _probably_ don't even occur. Still...

There's otherwise some tidying up and other minor changes.

Fixes zd# n/a

# Testing

How did you test?

Tested in Espressif environment and 

```
./configure --enable-all
make clean
make && make test
```

I did see a failure for this:

```
FAIL: testsuite/testsuite.test
```

But the exact same failure also occurs with a fresh clone of wolfssl.

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

SparkiDev commented on src/ssl.c:
```diff
@@ -1474,16 +1474,34 @@ WOLFSSL* wolfSSL_new(WOLFSSL_CTX* ctx)
             ssl = 0;
         }
-    }
+        else {
```
make this 
```
else if (ret == 0) {
...
}
else {
...
}
```

gojimmypi replied:
Agreed. Your suggestion results in improved readability. Done

SparkiDev commented on src/ssl.c:
```diff
@@ -1474,16 +1474,34 @@ WOLFSSL* wolfSSL_new(WOLFSSL_CTX* ctx)
+            }
+        } /* if InitSSL check */
+    } /* ssl XMALLOC success */
+    else {
```
Pull this up.
That is:
```
if (ssl != NULL) {
    WOLFSSL_MSG_EX("ssl xmalloc ...
}
else {
...
}
```

gojimmypi replied:
done

SparkiDev commented on src/ssl.c:
```diff
@@ -12598,10 +12625,13 @@ int wolfSSL_DTLS_SetCookieSecret(WOLFSSL* ssl,
+        }
         #endif
 
+        WOLFSSL_MSG(">> NOT TLS 1.3!\n");
```
"TLS 1.2 or lower"
will track better when reviewing logs.

gojimmypi replied:
Good. Done

SparkiDev commented on src/ssl.c:
```diff
@@ -12719,11 +12749,14 @@ int wolfSSL_DTLS_SetCookieSecret(WOLFSSL* ssl,
                     neededState = SERVER_HELLOVERIFYREQUEST_COMPLETE;
             #endif
             /* get response */
+            WOLFSSL_MSG("ssl.c ssl->options.serverState < neededState");
```
Don't put code into messages.
Describe what is going on.
Better would be:
"Server state up to needed state"

gojimmypi replied:
Agreed. Done.

SparkiDev commented on src/ssl.c:
```diff
@@ -12719,11 +12749,14 @@ int wolfSSL_DTLS_SetCookieSecret(WOLFSSL* ssl,
             /* get response */
+            WOLFSSL_MSG("ssl.c ssl->options.serverState < neededState");
             while (ssl->options.serverState < neededState) {
+                WOLFSSL_MSG("while (ssl->options.serverState < neededState)...");
```
same here
"Progressing server state" or something

gojimmypi replied:
much better, done.

SparkiDev commented on src/ssl_certman.c:
```diff
@@ -89,11 +89,20 @@ WOLFSSL_CERT_MANAGER* wolfSSL_CertManagerNew_ex(void* heap)
     WOLFSSL_CERT_MANAGER* cm;
 
     WOLFSSL_ENTER("wolfSSL_CertManagerNew");
+    if (heap == NULL) {
```
Why do we care?
%p will convert NULL to something different on at least some platforms.

gojimmypi replied:
> Why do we care?

I left a comment there for this. In my case _where_ the heap was being allocated was critical to the failure I was seeing. I had two segments of memory, and the heap was unexpectedly in the smaller one.

SparkiDev commented on src/tls.c:
```diff
@@ -7459,7 +7459,9 @@ static int TLSX_KeyShare_GenEccKey(WOLFSSL *ssl, KeyShareEntry* kse)
         kse->key = (byte*)XMALLOC(sizeof(ecc_key), ssl->heap, DYNAMIC_TYPE_ECC);
         if (kse->key == NULL) {
-            WOLFSSL_MSG("EccTempKey Memory error");
+            WOLFSSL_MSG_EX("Failed to allocate %d bytes, ssl->heap: %p",
```
Really necessary?

gojimmypi replied:
In memory constrained environments, I found this very valuable to see how much more memory was needed to avoid a failure.

I did however, leave the original `EccTempKey Memory error` just in case there was some sort of automated monitoring system that looked for that.

SparkiDev commented on src/tls13.c:
```diff
@@ -12620,7 +12620,7 @@ int wolfSSL_connect_TLSv13(WOLFSSL* ssl)
 
             ssl->options.connectState = CLIENT_HELLO_SENT;
-            WOLFSSL_MSG("connect state: CLIENT_HELLO_SENT");
+            WOLFSSL_MSG("tls13.c connect state: CLIENT_HELLO_SENT");
```
tls13.c => TLSv13

gojimmypi replied:
Agreed. Changed.

SparkiDev commented on wolfcrypt/src/aes.c:
```diff
@@ -530,7 +530,7 @@ block cipher mechanism that uses n-bit binary string parameter key with 128-bits
     #include <esp_log.h>
     #include <wolfssl/wolfcrypt/port/Espressif/esp32-crypt.h>
-    const char* TAG = "aes";
+    static const char* TAG = "aes";
```
Make it a #define.

SparkiDev commented on wolfcrypt/src/ecc.c:
```diff
@@ -3828,6 +3828,7 @@ int wc_ecc_mulmod_ex2(const mp_int* k, ecc_point* G, ecc_point* R, mp_int* a,
 #endif
    /* k can't have more bits than order */
    if (mp_count_bits(k) > mp_count_bits(order)) {
+      WOLFSSL_MSG("WARNING: mp_count_bits(k) > mp_count_bits(order)");
```
"Private key length is greater than order in bits."

gojimmypi replied:
Good suggestion. Done

SparkiDev commented on wolfcrypt/src/ecc.c:
```diff
@@ -5805,6 +5806,9 @@ static int _ecc_make_key_ex(WC_RNG* rng, int keysize, ecc_key* key,
         if (err == MP_OKAY) {
             err = wc_ecc_curve_load(key->dp, &curve, ECC_CURVE_FIELD_ALL);
         }
+        else {
```
Confusing, move up.

gojimmypi replied:
Agreed. Now that you mention it, that whole section was messy. I believe it is better now.

SparkiDev commented on wolfcrypt/src/ecc.c:
```diff
@@ -5815,7 +5819,9 @@ static int _ecc_make_key_ex(WC_RNG* rng, int keysize, ecc_key* key,
             err = ecc_make_pub_ex(key, curve, NULL, rng);
         }
-
+        else {
```
Confusing move up.

gojimmypi replied:
Same; I cleaned up that whole section regarding the verbose error messages.

SparkiDev commented on wolfcrypt/src/random.c:
```diff
@@ -1671,20 +1678,27 @@ static int _InitRng(WC_RNG* rng, byte* nonce, word32 nonceSz,
                 }
             }
-#else
+    #else
```
These aren't aligned with #ifdef???

gojimmypi replied:
I believe I fixed this one. Also added some whitespace around `CUSTOM_RAND_GENERATE_BLOCK` for clarity.

SparkiDev commented on wolfcrypt/src/random.c:
```diff
@@ -1671,20 +1678,27 @@ static int _InitRng(WC_RNG* rng, byte* nonce, word32 nonceSz,
                 ret = Hash_DRBG_Instantiate((DRBG_internal *)rng->drbg,
-                            seed + SEED_BLOCK_SZ, seedSz - SEED_BLOCK_SZ,
-                            nonce, nonceSz, rng->heap, devId);
+                    seed + SEED_BLOCK_SZ,
```
Don' spread them out over multiple lines.
Put it back.

gojimmypi replied:
Done (although I do believe it is more readable spread out; no worries).

SparkiDev commented on wolfcrypt/src/random.c:
```diff
@@ -1693,6 +1707,11 @@ static int _InitRng(WC_RNG* rng, byte* nonce, word32 nonceSz,
                 rng->drbg = NULL;
             }
         }
+        else {
```
Move up

gojimmypi replied:
Done

SparkiDev commented on wolfcrypt/src/random.c:
```diff
@@ -3744,15 +3766,23 @@ int wc_GenerateSeed(OS_Seed* os, byte* output, word32 sz)
     #ifndef NO_DEV_URANDOM /* way to disable use of /dev/urandom */
         os->fd = open("/dev/urandom", O_RDONLY);
+        #if defined(DEBUG_WOLFSSL)
+            WOLFSSL_MSG_EX("open(/dev/urandom... %d", 0);
```
No point in use %d

gojimmypi replied:
yes, my bad. fixed.

SparkiDev commented on wolfcrypt/src/random.c:
```diff
@@ -3744,15 +3766,23 @@ int wc_GenerateSeed(OS_Seed* os, byte* output, word32 sz)
     #endif
         {
+    #if defined(DEBUG_WOLFSSL)
+            WOLFSSL_MSG_EX("open(/dev/urandom... %d", 1);
```
No point in using %d
Do you mean /dev/random?

gojimmypi replied:
yes, my bad. fixed.

SparkiDev commented on wolfcrypt/src/rsa.c:
```diff
@@ -2681,12 +2681,16 @@ static int RsaFunctionSync(const byte* in, word32 inLen, byte* out,
 
-    if (INIT_MP_INT_SIZE(tmp, mp_bitsused(&key->n)) != MP_OKAY)
+    if (INIT_MP_INT_SIZE(tmp, mp_bitsused(&key->n)) != MP_OKAY) {
+        WOLFSSL_MSG("INIT_MP_INT_SIZE != MP_OKAY");
```
"Initialising MP to a size failed"

gojimmypi replied:
done

SparkiDev commented on wolfcrypt/src/rsa.c:
```diff
@@ -2704,14 +2708,18 @@ static int RsaFunctionSync(const byte* in, word32 inLen, byte* out,
-            if (mp_exptmod_nct(tmp, &key->e, &key->n, tmp) != MP_OKAY)
+            WOLFSSL_MSG("Calling mp_exptmod_nct...");
+            if (mp_exptmod_nct(tmp, &key->e, &key->n, tmp) != MP_OKAY) {
+                WOLFSSL_MSG("mp_exptmod_nct != MP_OKAY");
```
"!= MP_OKAY" => "failed"

gojimmypi replied:
done

SparkiDev commented on wolfcrypt/src/rsa.c:
```diff
@@ -2720,8 +2728,11 @@ static int RsaFunctionSync(const byte* in, word32 inLen, byte* out,
-             ret = MP_TO_E;
+        WOLFSSL_MSG("mp_to_unsigned_bin_len_ct...");
+        if (mp_to_unsigned_bin_len_ct(tmp, out, (int)*outLen) != MP_OKAY) {
+            WOLFSSL_MSG("mp_to_unsigned_bin_len_ct != MP_OKAY");
```
"!= MP_OKAY" => "failed"

gojimmypi replied:
done

SparkiDev commented on wolfcrypt/src/rsa.c:
```diff
@@ -2757,6 +2768,7 @@ static int wc_RsaFunctionSync(const byte* in, word32 inLen, byte* out,
 
     ret = wc_RsaEncryptSize(key);
     if (ret < 0) {
+        WOLFSSL_MSG("bad wc_RsaEncryptSize");
```
"wc_RsaEncryptSize failed"

gojimmypi replied:
done

SparkiDev commented on wolfcrypt/src/rsa.c:
```diff
@@ -3174,6 +3187,7 @@ static int wc_RsaFunction_ex(const byte* in, word32 inLen, byte* out,
     else
 #endif
     {
+        WOLFSSL_MSG("Calling wc_RsaFunctionSync...");
```
Why have this?

gojimmypi replied:
There was a large section of code where the embedded app would get stuck.

This was a breadcrumb, seemed useful at the time, but I've removed it.

SparkiDev commented on wolfcrypt/src/rsa.c:
```diff
@@ -3150,7 +3163,7 @@ static int wc_RsaFunction_ex(const byte* in, word32 inLen, byte* out,
     if (type == RSA_PRIVATE_DECRYPT &&
         key->state == RSA_STATE_DECRYPT_EXPTMOD) {
-
+        WOLFSSL_MSG("Calling RsaFunctionCheckIn...");
```
Why have this?

gojimmypi replied:
Same: It was a breadcrumb, seemed useful at the time, but I've removed it as I don't think it adds value at this time.

SparkiDev commented:
Don't want the debug to become too noisy.
That is, only put out debug statements on error or long operations or when required to debug why an error is happening.

gojimmypi commented:
> Don't want the debug to become too noisy.

Agreed. I removed some of the messages that were only breadcrumbs without any real value.

I hadn't intended this to be such a big code review, but I appreciate your time. I hope this update addresses all of your suggestions. All were completely valid. Thank you.

SparkiDev commented:
Thanks for all the changes!
